sns/parity_math.c

```diff
@@ -385,85 +768,911 @@ static void xor_calculate(struct m0_parity_math *math,
+	buf_sort(rs, total_count, math->pmi_data_count,
+		 fail, data, parity);
+
+	/* Sort failed buffer indices. */
+	fails_sort(rs, fail, total_count, math->pmi_parity_count);
+
+	VALUE_MISMATCH_ASSERT_INFO(fail_count, rs->rs_failed_nr);
+
+	/* Get encoding coefficient tables. */
+	ret = isal_gen_recov_coeff_tbl(math->pmi_data_count,
+				       math->pmi_parity_count, rs);
+	if (ret != 0)
+		return M0_ERR_INFO(ret, "failed to generate recovery "
+				   "coefficient tables");
+
+	/* Recover failed data. */
+	ec_encode_data(block_size, math->pmi_data_count,
+		       fail_count, rs->rs_decode_tbls,
+		       rs->rs_bufs_in, rs->rs_bufs_out);
+
+	return M0_RC(ret);
+}
+
+static void ir_rs_init(const struct m0_parity_math *math, struct m0_sns_ir *ir)
+{
+	M0_ENTRY("math=%p, ir=%p", math, ir);
+
+	M0_PRE(math != NULL);
+	M0_PRE(ir != NULL);
+
+	ir->si_rs.rs_encode_matrix = math->pmi_rs.rs_encode_matrix;
+	ir->si_rs.rs_decode_tbls = math->pmi_rs.rs_decode_tbls;
+	ir->si_rs.rs_failed_idx = math->pmi_rs.rs_failed_idx;
+	ir->si_rs.rs_alive_idx = math->pmi_rs.rs_alive_idx;
+
+	M0_LEAVE();
+}
+
+static void ir_rs_fini(struct m0_sns_ir *ir)
+{
+}
+
+static void ir_failure_register(struct m0_sns_ir *ir, uint32_t failed_index)
+{
+	M0_ENTRY("ir=%p, failed_index=%u", ir, failed_index);
+
+	M0_PRE(ir != NULL);
+	M0_PRE(ir->si_rs.rs_failed_idx != NULL);
+	M0_PRE(ir->si_rs.rs_failed_nr < ir->si_parity_nr);
+
+	ir->si_rs.rs_failed_idx[ir->si_rs.rs_failed_nr++] = failed_index;
+
+	M0_LEAVE();
+}
+
+static int ir_mat_compute(struct m0_sns_ir *ir)
+{
+	struct m0_sns_ir_block *blocks;
+	struct m0_reed_solomon *rs;
+	uint32_t                i;
+	uint32_t                alive_nr;
+	uint32_t                total_blocks_nr;
+	int                     ret;
+
+	M0_ENTRY("ir=%p", ir);
+	M0_PRE(ir != NULL);
+
+	blocks = ir->si_blocks;
+	rs = &ir->si_rs;
+	total_blocks_nr = ir_blocks_count(ir);
+
+	for (i = 0, alive_nr = 0; i < total_blocks_nr; i++)
+		if (blocks[i].sib_status == M0_SI_BLOCK_ALIVE)
+			rs->rs_alive_idx[alive_nr++] = blocks[i].sib_idx;
+
+	VALUE_MISMATCH_ASSERT_INFO(alive_nr, ir->si_alive_nr);
+
+	ret = isal_gen_recov_coeff_tbl(ir->si_data_nr, ir->si_parity_nr, rs);
+	if (ret != 0)
+		return M0_ERR_INFO(ret, "failed to generate decode matrix");
+
+	for (i = 0; i < rs->rs_failed_nr; i++)
+		dependency_bitmap_prepare(&blocks[rs->rs_failed_idx[i]], ir);
+
+	return M0_RC(ret);
+}
+
+static int ir_recover(struct m0_sns_ir *ir, struct m0_sns_ir_block *alive_block)
+{
+	struct m0_reed_solomon  *rs;
+	struct m0_bitmap        *failed_bitmap;
+	struct m0_bufvec        *alive_bufvec;
+	struct m0_bufvec        *failed_bufvec;
+	struct m0_buf            in_buf = M0_BUF_INIT0;
+	uint8_t                  curr_idx = UINT8_MAX;
+	uint32_t                 i;
+	uint32_t                 length;
+	int                      ret = 0;
+
+	M0_ENTRY("ir=%p, alive_block=%p", ir, alive_block);
+
+	rs = &ir->si_rs;
+
+	struct m0_buf            out_bufs[rs->rs_failed_nr];
```

Comment:
[Style] Move this declaration to the start only.

Reply:
Done.